Add workflow to synchronize go mod changes with submodules

codysoyland opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficient time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary

This PR attempt to improve the workflow of managing Dependabot updates to this repo. The problem is that when Dependabot updates a dependency in the main module, those updates are not reflected in the "submodule" (https://github.com/sigstore/sigstore-go/tree/main/examples/oci-image-verification) which includes a reference to the main module. That causes the build to fail, and one maintainer must manually run `go mod tidy` in the submodule and push those changes, and another maintainer must approve and merge them (as the last committer is disallowed from merging a PR). ([example](https://github.com/sigstore/sigstore-go/pull/311))

This PR automatically runs `go mod tidy` in both modules and commits/pushes them whenever Dependabot creates a new PR.

There is a caveat with this, which is that the actions do not run when the last committer used `GITHUB_TOKEN` to authenticate to the Git server, which is [intentional to avoid accidentally creating recursively invoked workflows](https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow). This can be worked around by using a PAT in this workflow to authenticate to the Git server, or maintainers can push an empty commit to trigger the Actions workflows.

If this seems like a good approach to fellow maintainers, I'd recommend we create a PAT for use within this repo and modify this workflow to use it.

This workflow could also be added to `sigstore/sigstore` for use with the KMS providers ([example](https://github.com/sigstore/sigstore/pull/1852)).

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

codysoyland committed:
Add workflow to synchronize go mod changes with submodules

Signed-off-by: Cody Soyland <codysoyland@github.com>

kommendorkapten reviewed (dismissed):
Thanks Cody! This looks good.
I think it may even work to just re-run the failing workflows from the UI instead of pushing an empty commit?

codysoyland commented:
> I think it may even work to just re-run the failing workflows from the UI instead of pushing an empty commit?

I don't think there's a way to trigger the workflows from the UI if they are not run due to the last commit originating from Actions -- they are not actually "failed" workflows, they just aren't created at all. Maybe I'm missing something, but I think it would require an empty commit to trigger the build, unless we add a PAT to a secret and use it to commit the go.mod changes.

@haydentherapper (or maybe @cpanato?), what do you think of incorporating this workflow and using a PAT to authenticate with git? As explained in the description, we need to use a PAT or the actions-generated commit will not trigger actions to run again, therefore the automation would be incomplete. We can create an org-level PAT secret and also use it to keep the kms providers in sigstore/sigstore up to date using the same workflow. Somebody who is org admin would need to create the PAT and upload it as an org-level secret.

Hayden-IO commented:
I really like this workflow, this is great! This has definitely been a nuisance here and in Sigstore/sigstore.

PATs are a little annoying to keep up to date, but we’ve already got a number that we manage, one more is fine. In the future, we can look at creating a GitHub app to replace PATs.  

what are the permissions needed for the PAT?

Hayden-IO commented on .github/workflows/synchronize-go-mod.yml:
```diff
@@ -0,0 +1,48 @@
+
+jobs:
+  synchronize-go-mod:
+    if: github.actor == 'dependabot[bot]'
```
Should this be an ID? If someone names their account this string, this would run. Not really a big deal though

codysoyland replied:
This is the [documented way](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions#responding-to-events). The actor name is a user name of the entity that triggered the workflow.

bobcallaway replied:
I agree with Hayden's question here - having this pinned to an immutable ID feels marginally safer.

Hayden-IO commented on .github/workflows/synchronize-go-mod.yml:
```diff
@@ -0,0 +1,48 @@
+      - name: Install Go
+        uses: actions/setup-go@0a12ed9d6a96ab950c8f026ed9f722fe0da7ef32 # v5.0.2
+        with:
+          go-version-file: go.mod
```
What if the modules don’t use the same version of go? Should we search for all go.mod files, then install go and run go mod tidy for each?

codysoyland replied:
I hadn't considered that, but I think we'd hit an error in that case, and we'd need to manually sync up the go versions.

kommendorkapten replied:
If we create a `go.work` and workspace, we can define the go version at the top level and use `go-version-file: go.work`, then rely on `go work sync` to propagate the version to the sub modules. That could probably be part of this workflow too.

codysoyland commented:
> PATs are a little annoying to keep up to date, but we’ve already got a number that we manage, one more is fine. In the future, we can look at creating a GitHub app to replace PATs.

Yeah an app is ideal, just requires a little more setup. PATs are indeed annoying.

> what are the permissions needed for the PAT?

It can be a fine-grained permission PAT with support to the specific repos we need (sigstore, sigstore-go), with "Contents" set to "Read and write".

Hayden-IO commented:
@cpanato  can you create the PAT?

Hayden-IO commented:
Coming back to this, I can create the PAT...but with that said, keeping these PATs up to date is a bit annoying as they expire. Is there any way to do this without another PAT?

codysoyland commented:
> Coming back to this, I can create the PAT...but with that said, keeping these PATs up to date is a bit annoying as they expire. Is there any way to do this without another PAT?

I believe we can do this with a GitHub app.... I can test it on my own org first to make sure and help you set it up if that works!

Hayden-IO commented:
We had also looked into https://github.com/apps/octo-sts awhile ago which Chainguard made.

cpanato commented:
sorry for the delay here

we can create a PAT, but i would vote to use the octo-sts, because we dont need to manage the PAT and that will be valid only for the job execution

cpanato commented:
and this looks soo cool!!

codysoyland commented:
octo-sts is awesome, I just tried it on a test repo and it works great for this purpose. I'll update the PR

Hayden-IO commented:
We'll need to install it in the org as well. Should mention this to the TSC before we progress.

codysoyland committed:
Use octo-sts instead of PAT

Signed-off-by: Cody Soyland <codysoyland@github.com>

codysoyland commented:
I updated the workflow to use octo-sts, so if TSC approves, we can proceed to merging this.

Hayden-IO commented:
@codysoyland Can you file an issue on https://github.com/sigstore/tsc asking for approval?

codysoyland commented:
> @codysoyland Can you file an issue on https://github.com/sigstore/tsc asking for approval?

https://github.com/sigstore/TSC/issues/51

bobcallaway commented on .github/workflows/synchronize-go-mod.yml:
```diff
@@ -0,0 +1,52 @@
+
+      - name: Synchronize go.mod in submodule
+        run: |
+          cd examples/oci-image-verification; go mod tidy
```
```suggestion
          cd examples/oci-image-verification && go mod tidy
```

Hayden-IO commented:
@codysoyland is this good to close?

codysoyland commented:
> @codysoyland is this good to close?

Yeah... the multi-directory config I added in #380 didn't help but this is resolved for grouped updates I added in #395. It's still an issue for major release updates, but I don't think it's worth dealing with octo-sts or PATs to make this workflow work.
